File: app/src/main/java/com/acdmt/task/TrialRunner.kt
package com.acdmt.task

import android.graphics.Color
import android.os.Handler
import android.os.Looper
import android.os.SystemClock
import com.acdmt.data.DataCollector
import com.acdmt.model.Hemifield
import com.acdmt.model.MatchDecision
import com.acdmt.model.StimulusItem
import com.acdmt.model.StimulusSlot
import com.acdmt.model.TrialPhase
import com.acdmt.model.TrialPlan
import com.acdmt.model.TrialRecord
import com.acdmt.render.StimulusRenderer
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withTimeoutOrNull
import kotlin.coroutines.resume
import kotlin.random.Random

class TrialRunner(
    private val renderer: StimulusRenderer,
    private val dataCollector: DataCollector,
    private val callbacks: Callbacks,
    private val random: Random = Random(System.currentTimeMillis())
) {
    interface Callbacks {
        fun onPhaseChanged(phase: TrialPhase, trialNumber: Int, setSize: Int)
        fun onResponseWindowChanged(isOpen: Boolean)
        fun onExperimentFinished(json: String)
    }

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    private val handler = Handler(Looper.getMainLooper())
    private var runJob: Job? = null
    private var currentPhase: TrialPhase = TrialPhase.IDLE
    private var currentSetSize = START_SET_SIZE
    private var consecutiveCorrect = 0
    private var responseDeferred: CompletableDeferred<ResponseEvent>? = null
    private var retrievalOnsetMs = 0L

    fun start() {
        if (runJob?.isActive == true) return

        dataCollector.clear()
        currentSetSize = START_SET_SIZE
        consecutiveCorrect = 0
        runJob = scope.launch {
            runExperiment()
        }
    }

    fun stop() {
        runJob?.cancel()
        responseDeferred?.cancel()
        handler.removeCallbacksAndMessages(null)
        callbacks.onResponseWindowChanged(false)
    }

    fun dispose() {
        stop()
        scope.cancel()
    }

    fun submitResponse(decision: MatchDecision) {
        if (decision == MatchDecision.NO_RESPONSE) return
        val pending = responseDeferred ?: return
        if (currentPhase != TrialPhase.RETRIEVAL || !pending.isActive) return

        val reactionTime = SystemClock.elapsedRealtime() - retrievalOnsetMs
        pending.complete(ResponseEvent(decision, reactionTime))
    }

    private suspend fun runExperiment() {
        for (trialNumber in 1..TOTAL_TRIALS) {
            val trial = createTrialPlan(trialNumber, currentSetSize)
            runTrial(trial)
        }

        transitionTo(TrialPhase.FINISHED, null)
        callbacks.onExperimentFinished(dataCollector.exportJsonString())
    }

    private suspend fun runTrial(trial: TrialPlan) {
        transitionTo(TrialPhase.ITI, trial)
        scheduledDelay(random.nextLong(ITI_MIN_MS, ITI_MAX_MS + 1))

        transitionTo(TrialPhase.FIXATION, trial)
        scheduledDelay(FIXATION_MS)

        transitionTo(TrialPhase.CUE, trial)
        scheduledDelay(CUE_MS)

        transitionTo(TrialPhase.ENCODING, trial)
        scheduledDelay(ENCODING_MS)

        transitionTo(TrialPhase.MAINTENANCE, trial)
        scheduledDelay(MAINTENANCE_MS)

        val response = runRetrieval(trial)
        val userDecision = response?.decision ?: MatchDecision.NO_RESPONSE
        val accuracy = if (isCorrect(userDecision, trial)) 1 else 0

        dataCollector.log(
            TrialRecord(
                trialNumber = trial.trialNumber,
                setSize = trial.setSize,
                cuedHemifield = trial.cuedHemifield,
                isMatchTrial = trial.isMatchTrial,
                userResponse = userDecision,
                accuracy = accuracy,
                reactionTimeMs = response?.reactionTimeMs
            )
        )

        updateStaircase(accuracy == 1)
    }

    private suspend fun runRetrieval(trial: TrialPlan): ResponseEvent? {
        responseDeferred = CompletableDeferred()
        retrievalOnsetMs = SystemClock.elapsedRealtime()
        transitionTo(TrialPhase.RETRIEVAL, trial)
        callbacks.onResponseWindowChanged(true)

        val response = withTimeoutOrNull(RETRIEVAL_MS) {
            responseDeferred?.await()
        }

        callbacks.onResponseWindowChanged(false)
        responseDeferred = null
        return response
    }

    private fun transitionTo(phase: TrialPhase, trial: TrialPlan?) {
        currentPhase = phase
        when (phase) {
            TrialPhase.ITI -> renderer.showBlank(phase)
            TrialPhase.FIXATION -> renderer.showFixation(phase)
            TrialPhase.CUE -> renderer.showCue(requireNotNull(trial).cuedHemifield)
            TrialPhase.ENCODING -> renderer.showEncoding(requireNotNull(trial))
            TrialPhase.MAINTENANCE -> renderer.showMaintenance()
            TrialPhase.RETRIEVAL -> renderer.showRetrieval(requireNotNull(trial))
            TrialPhase.FINISHED -> renderer.showFinished()
            TrialPhase.IDLE -> renderer.showBlank(phase)
        }
        callbacks.onPhaseChanged(phase, trial?.trialNumber ?: TOTAL_TRIALS, trial?.setSize ?: currentSetSize)
    }

    private suspend fun scheduledDelay(durationMs: Long) {
        suspendCancellableCoroutine { continuation ->
            val runnable = Runnable {
                if (continuation.isActive) continuation.resume(Unit)
            }
            handler.postDelayed(runnable, durationMs)
            continuation.invokeOnCancellation {
                handler.removeCallbacks(runnable)
            }
        }
    }

    private fun createTrialPlan(trialNumber: Int, setSize: Int): TrialPlan {
        val cuedHemifield = if (random.nextBoolean()) Hemifield.LEFT else Hemifield.RIGHT
        val isMatchTrial = random.nextBoolean()
        val leftItems = createHemifieldItems(Hemifield.LEFT, setSize)
        val rightItems = createHemifieldItems(Hemifield.RIGHT, setSize)
        val cuedItems = if (cuedHemifield == Hemifield.LEFT) leftItems else rightItems
        val uncuedItems = if (cuedHemifield == Hemifield.LEFT) rightItems else leftItems
        val cuedTestItems = if (isMatchTrial) {
            cuedItems.map { it.copy() }
        } else {
            createMismatchItems(cuedItems)
        }
        val testItems = if (cuedHemifield == Hemifield.LEFT) {
            cuedTestItems + uncuedItems.map { it.copy() }
        } else {
            uncuedItems.map { it.copy() } + cuedTestItems
        }

        return TrialPlan(
            trialNumber = trialNumber,
            setSize = setSize,
            cuedHemifield = cuedHemifield,
            isMatchTrial = isMatchTrial,
            memoryItems = leftItems + rightItems,
            testItems = testItems
        )
    }

    private fun createHemifieldItems(hemifield: Hemifield, setSize: Int): List<StimulusItem> {
        val selectedSlots = SLOT_TEMPLATE.shuffled(random).take(setSize)
        val selectedColors = COLOR_PALETTE.shuffled(random).take(setSize)
        return selectedSlots.zip(selectedColors).map { (slot, color) ->
            StimulusItem(
                hemifield = hemifield,
                slot = slot,
                color = color
            )
        }
    }

    private fun createMismatchItems(cuedItems: List<StimulusItem>): List<StimulusItem> {
        val changedIndex = random.nextInt(cuedItems.size)
        val usedColors = cuedItems.map { it.color }.toSet()
        return cuedItems.mapIndexed { index, item ->
            if (index != changedIndex) {
                item.copy()
            } else {
                // Exclude ALL colors already in this hemifield, not just the replaced item's color
                val replacement = COLOR_PALETTE
                    .filter { it !in usedColors }
                    .random(random)
                item.copy(color = replacement)
            }
        }
    }

    private fun isCorrect(decision: MatchDecision, trial: TrialPlan): Boolean {
        return when (decision) {
            MatchDecision.MATCH -> trial.isMatchTrial
            MatchDecision.MISMATCH -> !trial.isMatchTrial
            MatchDecision.NO_RESPONSE -> false
        }
    }

    private fun updateStaircase(wasCorrect: Boolean) {
        if (wasCorrect) {
            consecutiveCorrect += 1
            if (consecutiveCorrect >= 2) {
                currentSetSize = (currentSetSize + 1).coerceIn(MIN_SET_SIZE, MAX_SET_SIZE)
                consecutiveCorrect = 0
            }
        } else {
            consecutiveCorrect = 0
            currentSetSize = (currentSetSize - 1).coerceIn(MIN_SET_SIZE, MAX_SET_SIZE)
        }
    }

    private data class ResponseEvent(
        val decision: MatchDecision,
        val reactionTimeMs: Long
    )

    companion object {
        const val TOTAL_TRIALS = 100
        const val START_SET_SIZE = 2
        const val MIN_SET_SIZE = 3
        const val MAX_SET_SIZE = 8

        private const val ITI_MIN_MS  = 300L
        private const val ITI_MAX_MS  = 800L
        private const val FIXATION_MS = 500L
        private const val CUE_MS      = 300L
        private const val ENCODING_MS = 300L
        private const val MAINTENANCE_MS = 1_000L
        private const val RETRIEVAL_MS   = 2_000L

        private val SLOT_TEMPLATE = listOf(
            StimulusSlot(-0.82f, -0.58f),
            StimulusSlot(0f, -0.72f),
            StimulusSlot(0.82f, -0.58f),
            StimulusSlot(-0.82f, 0f),
            StimulusSlot(0.82f, 0f),
            StimulusSlot(-0.82f, 0.58f),
            StimulusSlot(0f, 0.72f),
            StimulusSlot(0.82f, 0.58f)
        )

        private val COLOR_PALETTE = listOf(
            Color.rgb(228, 30,  40),   // Red      H≈ 0°
            Color.rgb(242, 128, 20),   // Orange   H≈27°
            Color.rgb(242, 215, 20),   // Yellow   H≈53°
            Color.rgb(100, 222, 20),   // Lime     H≈93°
            Color.rgb(20,  188, 90),   // Green    H≈143°
            Color.rgb(20,  215, 222),  // Cyan     H≈182°
            Color.rgb(20,  100, 222),  // Blue     H≈216°
            Color.rgb(138, 20,  222),  // Purple   H≈276°
            Color.rgb(222, 20,  165),  // Magenta  H≈311°
            Color.WHITE                // White    achromatic
        )
    }
}

File: adaptive_wmapp_flutter/lib/services/trial_runner.dart
import 'dart:async';
import 'dart:math';
import 'package:flutter/material.dart';
import 'package:liblsl/lsl.dart';
import '../models/experiment_models.dart';
import 'data_collector.dart';
import 'edf_recorder.dart';

class TrialRunner extends ChangeNotifier {
  final DataCollector dataCollector;
  final EdfRecorder edfRecorder;
  final Random _random = Random();
  
  LSLStreamInfo? _streamInfo;
  LSLOutlet? _outlet;

  TrialPhase currentPhase = TrialPhase.idle;
  int currentSetSize = 2;
  int _consecutiveCorrect = 0;
  bool _isRunning = false;
  
  TrialPlan? currentTrial;
  Hemifield? currentCue;
  
  Completer<MatchDecision>? _responseCompleter;
  int _retrievalOnsetMs = 0;

  TrialRunner(this.dataCollector, this.edfRecorder) {
    _initLsl();
  }

  void _initLsl() async {
    try {
      _streamInfo = LSLStreamInfo(
        streamName: 'AdaptiveWM_Markers',
        streamType: LSLContentType.markers,
        channelCount: 1,
        sampleRate: 0.0,
        channelFormat: LSLChannelFormat.int32,
        sourceId: 'awm_uid'
      );
      _outlet = LSLOutlet(_streamInfo!);
      await _outlet!.create();
    } catch (e) {
      debugPrint("Failed to init LSL: $e");
    }
  }

  void _pushMarker(int marker) {
    try {
      if (_outlet != null) {
        _outlet!.pushSample([marker]);
      }
      edfRecorder.setMarker(marker);
    } catch (e) {
      debugPrint("Failed to push LSL marker $marker: $e");
    }
  }

  Future<void> start() async {
    if (_isRunning) return;
    _isRunning = true;
    dataCollector.clear();
    currentSetSize = 2;
    _consecutiveCorrect = 0;

    for (int trialNum = 1; trialNum <= 100; trialNum++) {
      if (!_isRunning) break;
      final trial = _createTrialPlan(trialNum, currentSetSize);
      await _runTrial(trial);
    }

    _transitionTo(TrialPhase.finished, null);
    await dataCollector.saveToCsvFile();
    _isRunning = false;
  }

  void stop() {
    _isRunning = false;
    _responseCompleter?.complete(MatchDecision.noResponse);
  }

  void submitResponse(MatchDecision decision) {
    if (decision == MatchDecision.noResponse) return;
    if (currentPhase != TrialPhase.retrieval || _responseCompleter == null || _responseCompleter!.isCompleted) return;
    
    _responseCompleter!.complete(decision);
  }

  Future<void> _runTrial(TrialPlan trial) async {
    _transitionTo(TrialPhase.iti, trial);
    await _delay(_random.nextInt(501) + 300); // 300 to 800

    _pushMarker(88); // Fixation marker
    _transitionTo(TrialPhase.fixation, trial);
    await _delay(500);

    _transitionTo(TrialPhase.cue, trial);
    await _delay(300);

    final encodingMarker = (trial.setSize * 10) + (trial.cuedHemifield == Hemifield.left ? 1 : 9);
    _pushMarker(encodingMarker);
    _transitionTo(TrialPhase.encoding, trial);
    await _delay(300);

    _transitionTo(TrialPhase.maintenance, trial);
    await _delay(1000);

    final response = await _runRetrieval(trial);
    final userDecision = response ?? MatchDecision.noResponse;
    
    int responseMarker = 12; // omission
    if (userDecision == MatchDecision.match) responseMarker = 11;
    if (userDecision == MatchDecision.mismatch) responseMarker = 10;
    _pushMarker(responseMarker);

    final isResponseCorrect = _isCorrect(userDecision, trial);
    final accuracy = isResponseCorrect ? 1 : 0;
    
    int? rt;
    if (userDecision != MatchDecision.noResponse) {
      rt = DateTime.now().millisecondsSinceEpoch - _retrievalOnsetMs;
    }

    dataCollector.log(
      TrialRecord(
        trialNumber: trial.trialNumber,
        setSize: trial.setSize,
        cuedHemifield: trial.cuedHemifield,
        isMatchTrial: trial.isMatchTrial,
        userResponse: userDecision,
        accuracy: accuracy,
        reactionTimeMs: rt,
      ),
    );

    _updateStaircase(isResponseCorrect);
  }

  Future<MatchDecision?> _runRetrieval(TrialPlan trial) async {
    _responseCompleter = Completer<MatchDecision>();
    _retrievalOnsetMs = DateTime.now().millisecondsSinceEpoch;
    _transitionTo(TrialPhase.retrieval, trial);

    try {
      final decision = await _responseCompleter!.future.timeout(const Duration(milliseconds: 2000));
      return decision;
    } on TimeoutException {
      return MatchDecision.noResponse;
    } finally {
      _responseCompleter = null;
    }
  }

  void _transitionTo(TrialPhase phase, TrialPlan? trial) {
    currentPhase = phase;
    currentTrial = trial;
    if (phase == TrialPhase.cue || phase == TrialPhase.encoding || phase == TrialPhase.retrieval) {
      currentCue = trial?.cuedHemifield;
    } else {
      currentCue = null;
    }
    notifyListeners();
  }

  Future<void> _delay(int milliseconds) async {
    if (!_isRunning) return;
    await Future.delayed(Duration(milliseconds: milliseconds));
  }

  TrialPlan _createTrialPlan(int trialNumber, int setSize) {
    final cuedHemifield = _random.nextBool() ? Hemifield.left : Hemifield.right;
    final isMatchTrial = _random.nextBool();
    
    final leftItems = _createHemifieldItems(Hemifield.left, setSize);
    final rightItems = _createHemifieldItems(Hemifield.right, setSize);
    
    final cuedItems = cuedHemifield == Hemifield.left ? leftItems : rightItems;
    final uncuedItems = cuedHemifield == Hemifield.left ? rightItems : leftItems;
    
    final cuedTestItems = isMatchTrial ? cuedItems.map((e) => e.copyWith()).toList() : _createMismatchItems(cuedItems);
    
    List<StimulusItem> testItems;
    if (cuedHemifield == Hemifield.left) {
      testItems = [...cuedTestItems, ...uncuedItems.map((e) => e.copyWith())];
    } else {
      testItems = [...uncuedItems.map((e) => e.copyWith()), ...cuedTestItems];
    }

    return TrialPlan(
      trialNumber: trialNumber,
      setSize: setSize,
      cuedHemifield: cuedHemifield,
      isMatchTrial: isMatchTrial,
      memoryItems: [...leftItems, ...rightItems],
      testItems: testItems,
    );
  }

  List<StimulusItem> _createHemifieldItems(Hemifield hemifield, int setSize) {
    final slots = List<StimulusSlot>.from(_slotTemplate)..shuffle(_random);
    final colors = List<Color>.from(_colorPalette)..shuffle(_random);
    
    final selectedSlots = slots.take(setSize).toList();
    final selectedColors = colors.take(setSize).toList();
    
    return List.generate(setSize, (i) => StimulusItem(
      hemifield: hemifield,
      slot: selectedSlots[i],
      color: selectedColors[i],
    ));
  }

  List<StimulusItem> _createMismatchItems(List<StimulusItem> cuedItems) {
    final changedIndex = _random.nextInt(cuedItems.length);
    final usedColors = cuedItems.map((e) => e.color).toSet();
    final availableColors = _colorPalette.where((c) => !usedColors.contains(c)).toList();
    final replacement = availableColors[_random.nextInt(availableColors.length)];
    
    final newItems = cuedItems.map((e) => e.copyWith()).toList();
    newItems[changedIndex] = newItems[changedIndex].copyWith(color: replacement);
    return newItems;
  }

  bool _isCorrect(MatchDecision decision, TrialPlan trial) {
    if (decision == MatchDecision.match) return trial.isMatchTrial;
    if (decision == MatchDecision.mismatch) return !trial.isMatchTrial;
    return false;
  }

  void _updateStaircase(bool wasCorrect) {
    if (wasCorrect) {
      _consecutiveCorrect++;
      if (_consecutiveCorrect >= 2) {
        currentSetSize = (currentSetSize + 1).clamp(3, 8);
        _consecutiveCorrect = 0;
      }
    } else {
      _consecutiveCorrect = 0;
      currentSetSize = (currentSetSize - 1).clamp(3, 8);
    }
  }

  @override
  void dispose() {
    _outlet?.destroy();
    _streamInfo?.destroy();
    super.dispose();
  }

  static const List<StimulusSlot> _slotTemplate = [
    StimulusSlot(-0.82, -0.58),
    StimulusSlot(0, -0.72),
    StimulusSlot(0.82, -0.58),
    StimulusSlot(-0.82, 0),
    StimulusSlot(0.82, 0),
    StimulusSlot(-0.82, 0.58),
    StimulusSlot(0, 0.72),
    StimulusSlot(0.82, 0.58)
  ];

  static const List<Color> _colorPalette = [
    Color.fromARGB(255, 228, 30, 40),
    Color.fromARGB(255, 242, 128, 20),
    Color.fromARGB(255, 242, 215, 20),
    Color.fromARGB(255, 100, 222, 20),
    Color.fromARGB(255, 20, 188, 90),
    Color.fromARGB(255, 20, 215, 222),
    Color.fromARGB(255, 20, 100, 222),
    Color.fromARGB(255, 138, 20, 222),
    Color.fromARGB(255, 222, 20, 165),
    Colors.white,
  ];
}
